addons/binding/org.openhab.binding.feican/src/main/java/org/openhab/binding/feican/internal/FeicanDiscoveryService.java

```diff
@@ -0,0 +1,196 @@
+            startDiscoverSocket();
+            // Runs until the socket call gets a time out and throws an exception. When a time out is triggered it means
+            // no data was present and nothing new to discover.
+            while (true) {
+                // Set packet length in case a previous call reduced the size.
+                receivePacket.setLength(buffer.length);
+                if (discoverSocket == null) {
+                    break;
+                } else {
+                    discoverSocket.receive(receivePacket);
+                }
+                logger.debug("Feican device discovery returned package with length {}", receivePacket.getLength());
+                if (receivePacket.getLength() > 0) {
+                    thingDiscovered(receivePacket);
+                }
+            }
+        } catch (SocketTimeoutException e) {
+            logger.debug("Discovering poller timeout...");
+        } catch (IOException e) {
+            logger.debug("Error during discovery: {}", e.getMessage());
+        } finally {
+            closeDiscoverSocket();
+            removeOlderResults(getTimestampOfLastScan());
+        }
+    }
+
+    /**
+     * Opens a {@link DatagramSocket} and sends a packet for discovery of Feican devices.
+     *
+     * @throws SocketException
+     * @throws IOException
+     */
+    private void startDiscoverSocket() throws SocketException, IOException {
+        SocketAddress sa = new InetSocketAddress(Connection.FEICAN_RECEIVE_PORT);
+        discoverSocket = new DatagramSocket(sa);
+        discoverSocket.setBroadcast(true);
+        discoverSocket.setSoTimeout(UDP_PACKET_TIMEOUT);
+        discoverSocket.send(discoverPacket);
+        if (logger.isTraceEnabled()) {
+            logger.trace("Discovery package sent: {}", new String(discoverPacket.getData(), StandardCharsets.UTF_8));
+        }
+    }
+
+    /**
+     * Closes the discovery socket and cleans the value. No need for synchronization as this method is called from a
+     * synchronized context.
+     */
+    private void closeDiscoverSocket() {
+        if (discoverSocket != null) {
+            discoverSocket.close();
+            discoverSocket = null;
+        }
+    }
+
+    /**
+     * Register a device (thing) with the discovered properties.
+     *
+     * @param packet containing data of detected device
+     */
+    private void thingDiscovered(DatagramPacket packet) {
+        final String ipAddress = packet.getAddress().getHostAddress();
+        if (packet.getData().length < 12) {
+            logger.debug(
+                    "Feican device was detected, but the retreived data is incomplete: '{}'. Device not registered",
+                    new String(packet.getData(), 0, packet.getLength() - 1, StandardCharsets.UTF_8));
+        } else {
+            String thingName = createThingName(packet.getData());
+            ThingUID thingUID = new ThingUID(THING_TYPE_BULB, thingName.toLowerCase(Locale.ENGLISH));
```

Comment:
Does thingName always conform to the restricted set of allowed chars for IDs?

Reply:
Did you consider this? IDs should only be `a-zA-Z_-`.

Reply:
The thing name will always be created as: `FC_<last for numbers of mac address>`. I think that is ok then?

Reply:
Yes!